### PR TITLE
File metadata

### PR DESCRIPTION
**Pull request by jpcima**
#330

This adds a metadata reader, which is capable of identifying sound files which contain wavetables.
This is determined by vendor-specific blocks which are present in the file metadata.

To implement this, a refactor was desired so the RIFF blocks are more easily accessible, and particularly in WAV files.

Wavetables Support
- Surge
- Serum
- u-he
- ~~some Native Instruments~~

EDIT removed Native, it was incorrect. these wavetables have their definition provided as sidecar text file.

### CODE REVIEW

**Comment by jpcima** [comment]
Added some tests to ensure wavetable is properly extracted.
- Surge: test file is made with `./scripts/wt-tool/add-surge-metadata.py`
- Clm ("Serum"): made with a custom code at `https://github.com/jpcima/WaveTableFactory/blob/627c20a7b35a4ab21dbb063eb0c5be35f0154acb/sources/wavetable.cpp#L71`
- U-he: too complex format, I don't know how to make (but we might create a dummy since it's just the chunk presence which is checked, not content)

File: src/sfizz/FilePool.cpp [hunk 1]
```diff
@@ -218,11 +218,12 @@ absl::optional<sfz::FileInformation> sfz::FilePool::getFileInformation(const Fil
...
     SF_INSTRUMENT instrumentInfo {};
 
-    const int sndFormat = reader->format();
-    if ((sndFormat & SF_FORMAT_TYPEMASK) == SF_FORMAT_FLAC)
-        sfz::FileInstruments::extractFromFlac(file, instrumentInfo);
-    else
-        reader->getInstrument(&instrumentInfo);
+    if (!reader->getInstrument(&instrumentInfo)) {
+        // if no instrument, then try extracting from embedded RIFF chunks (flac)
+        FileMetadataReader reader;
```
Comment (paulfd) [inline src/sfizz/FilePool.cpp]: This shadows the other reader

File: src/sfizz/FileMetadata.cpp [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+
+const RiffChunkInfo* FileMetadataReader::riffChunk(size_t index) const
+{
+    return impl_->riffChunk(index);
+}
+
+const RiffChunkInfo* FileMetadataReader::Impl::riffChunk(size_t index) const
+{
+    const std::vector<RiffChunkInfo>& riffChunks = riffChunks_;
+    return (index < riffChunks.size()) ? &riffChunks[index] : nullptr;
```
Comment (paulfd) [inline src/sfizz/FileMetadata.cpp]: I didn't try, but a static cast to `const RiffChunkInfo*` is not working?

**Review by paulfd — APPROVED** [review]

**Comment by paulfd** [comment]
I trust you for the file formats ;-)

File: src/sfizz/FileMetadata.cpp (hunk 2 above)
Comment (jpcima) [reply src/sfizz/FileMetadata.cpp]: Sorry explain what is the problem please, I don't understand it.

Comment (paulfd) [reply src/sfizz/FileMetadata.cpp]: Not really a problem per-say. I'm thinking you're creating the const ref to get a const pointer from the original vector, and I'm wondering if for some reason it's not possible to just cast to `const RiffChunkInfo*`. But it's not blocking, just curiosity. I will just try on my copy :D

Comment (paulfd) [reply src/sfizz/FileMetadata.cpp]:
```c++
const RiffChunkInfo* FileMetadataReader::Impl::riffChunk(size_t index) const
{
    return (index < riffChunks_.size()) ? &riffChunks_[index] : nullptr;
}
```
This seems to compile for me.

Comment (jpcima) [reply src/sfizz/FileMetadata.cpp]: Oh you prefer the one-line, sure I'll change it.

**Pull request merged**